File: probs/1068.c
#include <stdio.h>

int N, P[50], dp[50] = { 0, }, D, root;

int dfs(int n);

int main()
{
    freopen("input.txt", "rt", stdin);

    scanf("%d", &N);
    for (int i = 0; i < N; i++) {
        scanf("%d", &P[i]);
        if (P[i] == -1)
            root = i;
    }
    scanf("%d", &D);

    dfs(root);

    printf("%d\n", dp[root] - dp[D] + (D != root && dp[P[D]] == dp[D]));
}

int dfs(int n)
{
    int cnt = 0;
    for (int i = 0; i < N; i++)
        if (P[i] == n)
            cnt += dfs(i);
    if (cnt == 0) cnt = 1;

    dp[n] = cnt;
    return cnt;
}

File: probs/1068.py
import sys
input = sys.stdin.readline

input = open('input.txt', 'r').readline

def dfs(n):
    cnt = 0
    for i, p in enumerate(P):
        if p == n:
            cnt += dfs(i)
    if cnt == 0:
        cnt = 1
    dp[n] = cnt
    return cnt

N = int(input())
P = list(map(int, input().split()))
D = int(input())

root = P.index(-1)
dp = [0]*N
dfs(root)

print(dp[root] - dp[D] + (P[D] != -1 and dp[P[D]] == dp[D]))
